libraries/AP_Math/SCurve.cpp

```diff
@@ -0,0 +1,963 @@
+    // avoid re-calculating if unnecessary
+    if (is_equal(segment[SEG_INIT].end_vel, speed)) {
+        return speed;
+    }
+
+    const float Vm = segment[SEG_ACCEL_END].end_vel;
+    const float track_length = track.length();
+    speed = MIN(speed, Vm);
+
+    float Jm, t2, t4, t6;
+    calculate_path(jerk_time, jerk_max, speed, accel_max, Vm, track_length / 2.0f, Jm, t2, t4, t6);
+
+    uint8_t seg = SEG_INIT;
+    add_segment(seg, 0.0f, SegmentType::CONSTANT_JERK, 0.0f, 0.0f, speed, 0.0f);
+    add_segments_jerk(seg, jerk_time, Jm, t2);
+    add_segment_const_jerk(seg, t4, 0.0f);
+    add_segments_jerk(seg, jerk_time, -Jm, t6);
+
+    // remove numerical errors
+    segment[SEG_ACCEL_END].end_accel = 0.0f;
+
+    // offset acceleration segment if we can't fit it all into half the original length
+    const float dPstart = MIN(0.0f, track_length / 2.0f - segment[SEG_ACCEL_END].end_pos);
+    const float dt =  dPstart / segment[SEG_ACCEL_END].end_vel;
+    for (uint8_t i = SEG_INIT; i <= SEG_ACCEL_END; i++) {
+        segment[i].end_time += dt;
+        segment[i].end_pos += dPstart;
+    }
+
+    // add empty speed change segments and constant speed segment
+    for (uint8_t i = SEG_ACCEL_END+1; i <= SEG_CHANGE_END; i++) {
+        segment[i].seg_type = SegmentType::CONSTANT_JERK;
+        segment[i].jerk_ref = 0.0f;
+        segment[i].end_time = segment[SEG_ACCEL_END].end_time;
+        segment[i].end_accel = 0.0f;
+        segment[i].end_vel = segment[SEG_ACCEL_END].end_vel;
+        segment[i].end_pos = segment[SEG_ACCEL_END].end_pos;
+    }
+
+    seg = SEG_CONST;
+    add_segment_const_jerk(seg, 0.0f, 0.0f);
+
+    calculate_path(jerk_time, jerk_max, 0.0f, accel_max, segment[SEG_CONST].end_vel, track_length / 2.0f, Jm, t2, t4, t6);
+
+    add_segments_jerk(seg, jerk_time, -Jm, t6);
+    add_segment_const_jerk(seg, t4, 0.0f);
+    add_segments_jerk(seg, jerk_time, Jm, t2);
+
+    // remove numerical errors
+    segment[SEG_DECEL_END].end_accel = 0.0f;
+    segment[SEG_DECEL_END].end_vel = MAX(0.0f, segment[SEG_DECEL_END].end_vel);
+
+    // add to constant velocity segment to end at the correct position
+    const float dP = MAX(0.0f, track_length - segment[SEG_DECEL_END].end_pos);
+    const float t15 =  dP / segment[SEG_CONST].end_vel;
+    for (uint8_t i = SEG_CONST; i <= SEG_DECEL_END; i++) {
+        segment[i].end_time += t15;
+        segment[i].end_pos += dP;
+    }
+
+    // catch calculation errors
+    if (!valid()) {
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+        ::printf("SCurve::set_origin_speed_max invalid path\n");
+        debug();
+#endif
+        INTERNAL_ERROR(AP_InternalError::error_t::invalid_arg_or_result);
+        init();
+        return 0.0f;
+    }
+
+    return speed;
+}
+
+// set the maximum vehicle speed at the destination
+void SCurve::set_destination_speed_max(float speed)
+{
+    // if path is zero length then start speed must be zero
+    if (num_segs != segments_max) {
+        return;
```

Comment:
Shouldn't the caller be told?

Reply:
this is a bit of defensive programming but it shouldn't happen normally.  If it does happen then it means something has probably gone wrong somewhere else and the user will have been told in that other place.

I think we're perhaps starting to go overboard on the internal checks now so I'll leave this as-is.